### PR TITLE
Sync a users contacts birthday to the users birthday calendar

### LINKED ISSUE #21750: [CalDAV/CardDAV] birthday calendar
Question to ask:
- is the birthday calendar a server side feature?
- or is this something we want to build into the client (calendar-rework app)?

Seems like there are devices/apps out there which already extract birthday information from synced contacts. This is basically quite logic if you have all contacts at hand.

@georgehrke @Henni @irgendwie @raghunayyar @jancborchardt @karlitschek

### PR DESCRIPTION
**Pull request by DeepDiver1975**
fixes #21750

TODO:
- [x] unit tests
- [x] occ command to sync contacts and the birthday calendar
- [x] migration step to generate the birthday_calendar
# Test Scenario
1. add a contact with a birthday to any of a users addressbook
2. hook up the birthday calendar in your calendar client (or use http://github.com/owncloud/calendar-rework)
3. see the birthday being displayed

bildschirmfoto von 2016-02-18 14-54-29

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
> migration step to generate the birthday_calendar

Migration works implicit because in case we add a new contact the birthday service is call implicitly

**Comment by DeepDiver1975** [comment]
@georgehrke @jancborchardt @nickvergessen please test - THX

**Comment by rullzer** [comment]
> 1. add a contact with a birthday to any of a users addressbook

Should contactsrework be able to do this?

**Comment by DeepDiver1975** [comment]
> Should contactsrework be able to do this?

not yet - use e.g. evolution

**Comment by rullzer** [comment]
So for some reason I don't see this popping up in my calender app... but it does showup correctly on the dav endpoint!

:+1:

**Comment by nickvergessen** [comment]
Works here, however one question:
Does the admin have to call the occ command manually? or can we do this on update?

**Comment by nickvergessen** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
> Does the admin have to call the occ command manually? or can we do this on update?

On upgrade the birthday cal is generated implicit as contacts are being moved over - no additional migration step required.

The command is there as fall back.

File: apps/dav/lib/carddav/carddavbackend.php [hunk 1]
```diff
@@ -64,15 +66,22 @@ class CardDavBackend implements BackendInterface, SyncSupport {
...
 	 * CardDavBackend constructor.
 	 *
 	 * @param IDBConnection $db
 	 * @param Principal $principalBackend
+	 * @param EventDispatcherInterface $dispatcher
 	 */
-	public function __construct(IDBConnection $db, Principal $principalBackend) {
+	public function __construct(IDBConnection $db,
+								Principal $principalBackend,
+								$dispatcher ) {
```
Comment (nickvergessen) [inline apps/dav/lib/carddav/carddavbackend.php]: no typehint?

File: apps/dav/lib/caldav/birthdayservice.php [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	 * @param string $cardUri
+	 * @param string $cardData
+	 */
+	public function onCardChanged($addressBookId, $cardUri, $cardData) {
+
+		$book = $this->cardDavBackEnd->getAddressBookById($addressBookId);
+		$principalUri = $book['principaluri'];
+		$calendarUri = self::BIRTHDAY_CALENDAR_URI;
+		$calendar = $this->ensureCalendarExists($principalUri, $calendarUri, []);
+		$objectUri = $book['uri'] . '-' . $cardUri. '.ics';
```
Comment (DeepDiver1975) [inline apps/dav/lib/caldav/birthdayservice.php]: I was thinking about using a different uri naming schema - simply extract the uid from the vcard and use that ....

Comment (nickvergessen) [inline apps/dav/lib/caldav/birthdayservice.php]: Well could be duplicate across addressbooks?

Comment (rullzer) [inline apps/dav/lib/caldav/birthdayservice.php]: does it really matter... the current one is guaranteed to be unique... better to be on the safe side i'd say...

File: apps/dav/lib/carddav/carddavbackend.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/dav/lib/carddav/carddavbackend.php]: no because I want to be able to pass in null

Comment (nickvergessen) [inline apps/dav/lib/carddav/carddavbackend.php]: So `EventDispatcherInterface $dispatcher = null` should do it while preventing inserting :hankey:

**Pull request merged**